compiler/rustc_lint/src/builtin.rs

```diff
@@ -1839,6 +1839,72 @@ impl EarlyLintPass for IncompleteFeatures {
     }
 }
 
+declare_lint! {
+    pub INCOMPATIBLE_FEATURES,
+    Deny,
+    "Incompatible features that would result in an undefined behavior if present at the same time"
+}
+
+declare_lint_pass!(
+    /// Check for used feature gates in `INCOMPATIBLE_FEATURES` in `librustc_feature/active.rs`.
+    IncompatibleFeatures => [INCOMPATIBLE_FEATURES]
+);
+
+impl EarlyLintPass for IncompatibleFeatures {
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, _: &ast::Crate) {
+        // track which features we have visited from the incompatible list
+        let mut visited = [(false, false); rustc_feature::INCOMPATIBLE_FEATURES.len()];
+
+        let features = cx.sess.features_untracked();
+        let filtered_features: Vec<(&Symbol, &Span)> = features
+            .declared_lang_features
+            .iter()
+            .map(|(name, span, _)| (name, span))
+            .chain(features.declared_lib_features.iter().map(|(name, span)| (name, span)))
+            .filter(|(name, _)| {
+                rustc_feature::INCOMPATIBLE_FEATURES.iter().enumerate().any(|(i, (f1, f2))| {
+                    // mark which are visited accordingly and keep it, otherwise filter it
+                    if name == &f1 {
+                        visited[i].0 = true;
+                        true
+                    } else if name == &f2 {
+                        visited[i].1 = true;
+                        true
+                    } else {
+                        false
+                    }
+                })
+            })
+            .collect();
+
+        visited
+            .iter()
+            .enumerate()
+            .filter(|(_, (f1, f2))| *f1 && *f2) // both are found
+            .map(|(i, _)| rustc_feature::INCOMPATIBLE_FEATURES[i])
+            .for_each(|(ref f1, ref f2)| {
+                if let Some((f1_name, f1_span)) =
+                    filtered_features.iter().find(|(name, _)| name == &f1)
+                {
+                    if let Some((f2_name, f2_span)) =
+                        filtered_features.iter().find(|(name, _)| name == &f2)
+                    {
+                        let spans = vec![**f1_span, **f2_span];
+                        cx.struct_span_lint(INCOMPATIBLE_FEATURES, spans.clone(), |lint| {
+                            lint.build(&format!(
+                                "features `{}` and `{}` are incompatible, using them \
+                                at the same time can result in an undefined behavior",
```

Comment:
```suggestion
                                at the same time can result in undefined behavior",
```

Reply:
Undefined behavior has a very specific meaning: the compiler is allowed to do anything to your code, up to and including segfaults, security vulnerabilities, and just giving an error. Is that's what's intended here? Or is this just saying 'rustc could be buggy'? I think we should distinguish between compiler _bugs_ and behavior that is _intended_ to be undefined.

cc @RalfJung

Reply:
Its not that 'rustc is buggy', to fix the issue we might have checked everytime `min_const_generics` is used that `const_generics` is not present, but I thought of making this general in case in the future we encountered something similar.


I intended to produce an error so that users cannot use them together, It thought rustc_lint was the correct place.

Reply:
Ok, in that case let's just say 'not allowed'. UB has other connotations which aren't applicable here.

Reply:
I think it's better to produce an error instead of the current warning. Do you know how to do that?

Reply:
I think you can just change this call from `struct_span_lint` to `struct_span_err`.

Reply:
UB would be accurate here if there are known soundness bugs -- like there are for incomplete features, though I think there I made the text say something like "unsafe to use" or so.

Reply:
Oh, but this seems different. We just want people to disable `min_const_generic` when enabling `const_generic`. And this is a hard error, not a warning. So yeah, it probably doesn't make sense to speak about UB here.

Reply:
yah, working on making it hard error

Reply:
Should I still do it in `rustc_lint`? or which part of `rustc` is better for this?

Reply:
I think `rustc_ast_passes/feature_gate.rs` might be fitting, not too sure about that though